### PR TITLE
Make CRUDHandler generic (attempt 2)

### LINKED ISSUE #5555: Make CRUDHandler generic to its resource
Currently the entire typing of the CRUDHandler (and normal handler) is to the Resource type. However, in handler we often override this to the actual resource. However, type wise this is not correct.

The goal of this ticket it to make the CRUDHandler generic (in a non breaking way). For example:

```python
class LicenseResource(resources.Resource, base.ApiBase):
    fields = ...


class LicenseHandler(handler.CRUDHandler[LicenseResource]):
    def read_resource(self, ctx: handler.HandlerContext, resource: LicenseResource) -> None: ...
```

### PR DESCRIPTION
**Pull request by arnaudsjs**
# Description

This PR adds the CRUDHandlerGeneric class. This class contains the same implementation as the CRUDHandler class, but is generic with respect to the specific PurgeableResource it acts on.

closes #5555 and #6146

# Self Check:

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
- [x] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )
- [ ] ~~If this PR fixes a race condition in the test suite, also push the fix to the relevant stable branche(s) (see [test-fixes](https://internal.inmanta.com/development/core/tasks/build-master.html#test-fixes) for more info)~~

### CODE REVIEW

File: src/inmanta/agent/handler.py [hunk 1]
```diff
@@ -932,6 +933,14 @@ def execute(self, ctx: HandlerContext, resource: resources.Resource, dry_run: Op
                 )
 
 
+# Keep the CRUDHandler class for backwards compatibility. Making the CRUDHandler class generic can cause MRO issues
+# if a subclass of CRUDHandler is generic as well. The code below makes a copy of the CRUDHandlerGeneric class
+# and removes Generic as a base class.
+CRUDHandler = types.new_class(
```
Comment (sanderr) [inline src/inmanta/agent/handler.py:939]: Iirc my testing revealed that it isn't a problem per se that it inherits from `Generic`, more that it inherited from a non-parametrized generic. I think we should be fine with
```python
class CRUDHandler(CRUDHandlerGeneric[resources.PurgeableResource]): pass

**Review by sanderr — CHANGES_REQUESTED** [review]

File: src/inmanta/agent/handler.py (hunk 1 above)
Comment (arnaudsjs) [reply src/inmanta/agent/handler.py:939]: No. The issue is the method resolution order (MRO) that cannot be determined. The inheritance structure looks like this:

```mermaid
classDiagram
    class Generic
    class CRUDHadlerGeneric
    class CRUDHandler
    class BaseHandler
    class CFSOrderResourceHandler

    Generic <|-- CRUDHadlerGeneric
    CRUDHadlerGeneric <|-- CRUDHandler
    CRUDHandler <|-- BaseHandler
    BaseHandler <|-- CFSOrderResourceHandler
    Generic <|-- CFSOrderResourceHandler
```

The `CFSOrderResourceHandler` class is defined as:

```
class CFSOrderResourceHandler(Generic[Order], BaseHandler):
```

which makes it impossible to find a consistent ordering. If the signature would be changed to:

```
class CFSOrderResourceHandler(BaseHandler, Generic[Order])
```

it works. But that is not going to help us here.

I tested with your proposal and the MRO issue is still there.

**Review by sanderr — COMMENTED** [review]
As discussed

File: src/inmanta/agent/handler.py [hunk 2]
```diff
@@ -952,6 +952,14 @@ def delete_resource(self, ctx: HandlerContext, resource: TPurgeableResource) ->
     def update_resource(self, ctx: HandlerContext, changes: Dict[str, Dict[str, Any]], resource: TPurgeableResource) -> None:
         pass
 
+    def calculate_diff(
+        self, ctx: HandlerContext, current: TPurgeableResource, desired: TPurgeableResource
+    ) -> typing.Dict[str, typing.Dict[str, typing.Any]]:
+        super().calculate_diff(ctx, current, desired)
```
Comment (sanderr) [inline src/inmanta/agent/handler.py:958]: Could just use `typing.overload` as @wouterdb proposed. But I don't think this one additional super call should have any real effect on performance in practice.

**Review by sanderr — APPROVED** [review]

**Pull request closed without merging**